doc/reviewing-contributions: Add points about patches

roberth opened the pull request:
###### Description of changes

A weaker version of https://github.com/NixOS/nixpkgs/pull/186484#issuecomment-1608960696 
Does not include a consideration of whether patches should be packaged at all.
I believed that this PR (as is) merely documents existing standards whereas the linked comment would suggest a policy discussion that can be had, but I'd rather documented the baseline first. In other words I'd like to scope that out of this PR's discussion, so that we can make incremental progress.

EDIT: believed, past tense.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

roberth committed:
doc/reviewing-contributions: Add points about patches

A weaker version of https://github.com/NixOS/nixpkgs/pull/186484#issuecomment-1608960696

alyssais commented on doc/contributing/reviewing-contributions.chapter.md:
```diff
@@ -62,6 +62,8 @@ Sample template for a package update review is provided below.
 - [ ] executables tested on ARCHITECTURE
 - [ ] all depending packages build
+- [ ] patches have a comment describing either the upstream URL or a reason why the patch wasn't upstreamed
+- [ ] patches that are remotely available are fetched with `fetchpatch`
```
I would like to either not mention how patches should be fetched, or explain the nuances of when to use `fetchpatch` vs `fetchurl`, because I see a lot of misunderstanding about it.  (Maybe this would make more sense elsewhere in the manual, and we could link to it here.)

`fetchpatch` for when a patch is generated in response to the request for it — .patch URLs provided by GitHub, for example.  It normalises the patches so that it (hopefully) won't have a different hash the next time it's fetched, which is good, but that comes at the consequence of complexity.  One of the main downsides to `fetchpatch` is that it depends on a specific, old version of patchutils to do this normalization, and because we can't ever change that, we now have `fetchpatch2`, which uses a newer version.  Additionally, some packages that are early enough in the bootstrap chain can't use `fetchpatch`, because they are dependencies of patchutils.

So for patches that are static files (e.g. patches checked into another distro's git repo, or raw messages from a mailing list archive), it's preferable to use `fetchurl` so we aren't adding even more dependencies on the behaviour of a specific version of patchutils when we don't need to.

roberth replied:
Indeed this doesn't seem to have been documented yet. I've split that into a separate issue, #240141, because it's a related pre-existing problem.

alyssais commented:
> I believe that this PR (as is) merely documents existing standards

I don't think this is quite true, as I've not seen much insistence so far on comments explaining why a patch couldn't be upstreamed, but I think that would be a worthy thing to adopt (and a small, easy change to make), so no arguments from me there.

roberth commented on doc/contributing/reviewing-contributions.chapter.md:
```diff
@@ -105,6 +107,7 @@ Sample template for a new package review is provided below.
 - [ ] the list of `phases` is not overridden
 - [ ] when a phase (like `installPhase`) is overridden it starts with `runHook preInstall` and ends with `runHook postInstall`.
+- [ ] patches have a comment describing either the upstream URL or a reason why the patch wasn't upstreamed
 - [ ] patches that are remotely available are fetched with `fetchpatch`
```
```suggestion
- [ ] patches that are remotely available are fetched rather than vendored
```

roberth committed:
doc/reviewing-contributions: Remove wrong fetchpatch detail

It may have to be fetchpatch, fetchpatch2, or fetchurl. We should link somewhere, but we don't have a link target yet, so until then, remove the inaccurate info.

Co-authored-by: Alyssa Ross <hi@alyssa.is>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/require-a-comment-for-each-patch/29705/1
